File: src/yashigani/backoffice/routes/envelope_reapproval.py
"""
Yashigani Backoffice — Capability-Envelope RE-APPROVAL admin routes (3.0).

The operator-facing admin surface for the imported-MCP capability-envelope pin
(YSG-RISK-060).  Tom built the backend — the durable envelope ledger
(``mcp/envelope_service.py``), the field-level diff + step-up re-approve
(``mcp/envelope_reapproval.py``), and the refresh triage
(``mcp/_envelope_triage.py``).  This route is the THIN API seam the
re-approval SPA consumes; it wires the REAL services (no stub) and surfaces:

  GET    /admin/mcp/envelopes/pending                  — pending re-approval queue
  GET    /admin/mcp/envelopes/pending/{prov}           — field-level diff
                                                          (candidate vs ORIGINAL
                                                          baseline AND vs prior)
  POST   /admin/mcp/envelopes/pending/{prov}/approve   — step-up re-approve →
                                                          re-pin new baseline
  POST   /admin/mcp/envelopes/pending/{prov}/reject    — keep-blocked (step-up)

The candidate (refreshed) surface that triggered a block is held by the
backoffice-owned :class:`EnvelopePendingStore` (Redis db/3), written by the
broker when it latches the block.  The diff is ALWAYS recomputed server-side vs
the ORIGINAL baseline (envelope_version 1) — the anti-rug-pull framing — using
the same deterministic structural diff the broker uses; the store's pre-rendered
findings are advisory only.

Security properties enforced here (the brief's QA mandate on our own build):
  * **RBAC + step-up on approve/reject** — ``reapprove_envelope`` calls the
    shared ``assert_privileged_mutation`` gate (operator tier + FRESH TOTP),
    which raises ``StepUpRequired`` (401, ``error=step_up_required``) so the JS
    interceptor prompts for TOTP, then retries.  A non-admin gets 403; a hijacked
    session without the second factor cannot re-pin or clear a queue entry.
  * **Tenant scoping (BOLA close)** — every read/mutation is scoped to this
    install's tenant; a pending entry from another tenant is invisible and a
    cross-tenant approve/reject fails closed (404).
  * **No secret / no handle in the response** — only the public, typed
    capability dimensions + diff details are returned.  No envelope handle, no
    map material, no salt.
  * **Output escaping is the UI's job** — tool keys, descriptions and diff
    ``detail`` strings derive from an UNTRUSTED upstream MCP (the XSS surface).
    This route returns them as JSON strings (no HTML); the renderer
    (envelope_reapproval.js) escapeHtml()s every field at the DOM sink.

# Last updated: 2026-06-10
"""
from __future__ import annotations

import logging
import os

from fastapi import APIRouter, HTTPException, status

from yashigani.backoffice.middleware import AdminSession
from yashigani.backoffice.state import backoffice_state
from yashigani.common.error_envelope import safe_error_envelope

logger = logging.getLogger(__name__)

router = APIRouter()


# ── Tenant / service resolution ────────────────────────────────────────────

def _install_tenant() -> str:
    """Resolve this install's tenant id (mirrors documents._install_tenant).

    Single-tenant installs use the stable default ``"default"``; a multi-tenant
    deployment sets ``YASHIGANI_TENANT_ID`` per install.  Every read + mutation
    here is scoped to this tenant so a pending re-approval (and its candidate)
    never crosses a tenant boundary (BOLA close)."""
    return os.environ.get("YASHIGANI_TENANT_ID", "default").strip() or "default"


def _pending_store():
    """Return the wired EnvelopePendingStore or 503 (fail-closed).

    A read/mutation must never appear to succeed against a store that does not
    exist — the operator would see an empty queue and assume nothing is blocked
    when in fact the store is simply down."""
    store = backoffice_state.envelope_pending_store
    if store is None:
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail={
                "error": "envelope_pending_store_unavailable",
                "message": "Capability-envelope pending store not initialised (Redis db/3 unavailable).",
            },
        )
    return store


def _envelope_service():
    """Construct a CapabilityEnvelopeService over the live asyncpg pool, or 503.

    The pool is opened in the backoffice lifespan after create_pool(); in
    dev/test without a DB it is absent → 503 (fail-closed) rather than a 500."""
    try:
        from yashigani.db import get_pool
        from yashigani.mcp.envelope_service import CapabilityEnvelopeService
        return CapabilityEnvelopeService(get_pool())
    except Exception as exc:  # noqa: BLE001
        logger.warning("envelope re-approval: service unavailable (%s)", exc)
        raise HTTPException(
            status_code=status.HTTP_503_SERVICE_UNAVAILABLE,
            detail={
                "error": "envelope_service_unavailable",
                "message": "Capability-envelope durable store not initialised (DB pool unavailable).",
            },
        )


# ── Diff rendering ─────────────────────────────────────────────────────────

# The diff dimensions, mapped to a human label + the "what an attacker is
# trying to gain" framing the SPA surfaces.  Severity drives the badge colour.
_DIMENSION_META = {
    "tool_set": {"label": "New tool", "severity": "high"},
    "effect_class": {"label": "Effect class", "severity": "high"},
    "egress": {"label": "Egress posture", "severity": "high"},
    "data_scope": {"label": "Data scope", "severity": "med"},
    "arg_shape": {"label": "Argument schema", "severity": "med"},
    "annotation": {"label": "Model-facing hint", "severity": "med"},
    "output": {"label": "Output shape", "severity": "med"},
    "unknown": {"label": "Unmodelled capability", "severity": "high"},
}


def _render_findings(findings: list) -> list:
    """Render a list[DiffFinding] into JSON-safe rows for the SPA.

    Every string here (tool_key / detail) is attacker-controlled — the route
    passes it through as JSON; the SPA escapeHtml()s it at the DOM sink."""
    out = []
    for f in findings:
        dim = getattr(f, "dimension", "") or ""
        meta = _DIMENSION_META.get(dim, {"label": dim or "change", "severity": "med"})
        out.append({
            "dimension": dim,
            "label": meta["label"],
            "severity": meta["severity"],
            "tool_key": getattr(f, "tool_key", "") or "",
            "detail": getattr(f, "detail", "") or "",
        })
    return out


def _envelope_view(env) -> dict:
    """A compact, JSON-safe view of a ServerEnvelope for the diff panel header.

    Surfaces the typed dimensions only (no secret, no hash material beyond the
    informational surface_set_hash).  Tool keys are untrusted → UI escapes."""
    tools = []
    for tk, t in sorted(env.tools.items()):
        tools.append({
            "tool_key": tk,
            "effect_classes": sorted(e.value for e in t.effect_classes),
            "data_scopes": sorted(t.data_scopes),
            "output_open": bool(t.output_open),
        })
    return {
        "egress_posture": env.egress_posture,
        "tool_count": len(env.tools),
        "tools": tools,
    }


# ── Pending queue ──────────────────────────────────────────────────────────

@router.get("/pending")
async def list_pending(session: AdminSession):
    """List imported MCPs whose tool-surface refresh is BLOCKED + awaiting an
    operator re-approval decision (this tenant only).

    Always 200 (empty queue renders the all-clear state)."""
    store = _pending_store()
    return {"pending": store.list_for_tenant(_install_tenant())}


@router.get("/pending/{provenance_id}")
async def get_pending_diff(provenance_id: str, session: AdminSession):
    """Field-level diff for ONE blocked refresh.

    Returns the candidate diffed vs the ORIGINAL approved baseline (cumulative
    drift since import — the anti-rug-pull framing) AND vs the immediately-prior
    approved envelope.  The diff is recomputed server-side with the authoritative
    deterministic structural diff; the operator approves the ABSOLUTE candidate,
    never a delta.

    404 if the provenance has no pending entry for this tenant (BOLA close)."""
    store = _pending_store()
    tenant = _install_tenant()

    row = store.get(provenance_id, tenant)
    if row is None:
        raise HTTPException(status_code=404, detail={"error": "no_pending_reapproval"})

    candidate = store.get_candidate_envelope(provenance_id, tenant)
    if candidate is None:  # pragma: no cover - get() already gated; defensive
        raise HTTPException(status_code=404, detail={"error": "no_pending_reapproval"})

    svc = _envelope_service()
    try:
        baseline = await svc.get_baseline_envelope(provenance_id)
        active = await svc.get_active_envelope(provenance_id)
    except Exception as exc:
        envelope, _ = safe_error_envelope(exc, public_message="envelope lookup failed")
        raise HTTPException(status_code=503, detail=envelope)

    if baseline is None:
        # No original baseline ⇒ never imported through the ceremony.  The
        # pending entry is stale/orphaned — surface it honestly, do not pretend.
        raise HTTPException(
            status_code=409,
            detail={"error": "no_baseline_envelope",
                    "message": "No original approved baseline for this MCP — cannot diff/re-approve."},
        )

    from yashigani.mcp.envelope_reapproval import compute_field_level_diff

    # vs PRIOR: the active row's typed dimensions == the baseline's until a
    # re-approval mints a new version, so when there is no separate active row we
    # fall back to the baseline (the diff vs prior then equals vs original).
    prior_env = active.envelope if active is not None else baseline.envelope
    fld = compute_field_level_diff(
        provenance_id=provenance_id,
        original_baseline=baseline.envelope,
        prior_envelope=prior_env,
        candidate=candidate,
    )

    return {
        "provenance_id": provenance_id,
        "server_id": row.get("server_id", ""),
        "triage_class": row.get("triage_class", ""),
        "blocked_at": row.get("blocked_at", 0.0),
        # Header views (typed dimensions only — no secrets).
        "original": _envelope_view(baseline.envelope),
        "candidate": _envelope_view(candidate),
        # The two diffs — vs ORIGINAL is the anti-rug-pull anchor.
        "vs_original": _render_findings(fld.vs_original),
        "vs_prior": _render_findings(fld.vs_prior),
        "egress_change": (
            baseline.envelope.egress_posture != candidate.egress_posture
        ),
        "egress_from": baseline.envelope.egress_posture,
        "egress_to": candidate.egress_posture,
    }


# ── Approve (step-up re-approve → re-pin new baseline) ──────────────────────

@router.post("/pending/{provenance_id}/approve")
async def approve_pending(provenance_id: str, session: AdminSession):
    """Re-approve a blocked tool-surface mutation behind the step-up gate.

    Wires the REAL ``reapprove_envelope`` — which calls the shared
    ``assert_privileged_mutation`` gate (operator RBAC + FRESH TOTP).  On a
    session without a fresh step-up it raises ``StepUpRequired`` (401,
    ``error=step_up_required``) so the JS interceptor prompts for TOTP and
    retries.  On success it mints a NEW chained envelope version (re-pins the new
    baseline) and CLEARS the latched block.  The pending entry is then consumed.

    A non-admin session is rejected (403) by the gate before any mutation."""
    store = _pending_store()
    tenant = _install_tenant()

    row = store.get(provenance_id, tenant)
    if row is None:
        raise HTTPException(status_code=404, detail={"error": "no_pending_reapproval"})
    candidate = store.get_candidate_envelope(provenance_id, tenant)
    if candidate is None:  # pragma: no cover - defensive
        raise HTTPException(status_code=404, detail={"error": "no_pending_reapproval"})

    svc = _envelope_service()
    from yashigani.mcp.envelope_reapproval import reapprove_envelope

    # reapprove_envelope enforces the step-up gate itself (StepUpRequired /
    # NotAuthorisedForPrivilegedMutation propagate as 401/403 to the client) and
    # mints the new baseline.  We do NOT pre-gate with StepUpAdminSession — that
    # would double-enforce and the gate inside reapprove is the audited one.
    try:
        new_id = await reapprove_envelope(
            session=session,
            envelope_service=svc,
            candidate=candidate,
            server_id=row.get("server_id", ""),
            operator_identity=session.account_id,
            topology="ring_fenced",
            audit_writer=backoffice_state.audit_writer,
        )
    except HTTPException:
        # StepUpRequired (401) / NotAuthorised (403) — propagate verbatim so the
        # JS interceptor sees error=step_up_required and prompts for TOTP.
        raise
    except Exception as exc:
        envelope, _ = safe_error_envelope(exc, public_message="re-approval failed")
        raise HTTPException(status_code=500, detail=envelope)

    # Mutation succeeded + audited — consume the pending entry.
    store.resolve(provenance_id, tenant)
    logger.warning(
        "capability-envelope RE-APPROVED via admin UI: provenance=%.12s new_id=%d by=%s",
        provenance_id, new_id, session.account_id,
    )
    return {
        "status": "ok",
        "provenance_id": provenance_id,
        "new_envelope_id": new_id,
        "message": "Capability envelope re-pinned to the new baseline; block cleared.",
    }


# ── Reject / keep-blocked (step-up) ────────────────────────────────────────

@router.post("/pending/{provenance_id}/reject")
async def reject_pending(provenance_id: str, session: AdminSession):
    """Reject the mutation — keep the MCP blocked.

    Also a privileged decision (a hijacked session must not silently clear the
    operator's re-approval queue), so it is gated by the same
    ``assert_privileged_mutation`` step-up.  On reject the DB block STAYS latched
    (no new baseline is minted — only a step-up re-approval can clear it); the
    pending queue entry is consumed so it no longer nags the operator.  The
    upstream MCP remains hard-gated at the invocation broker (fail-closed)."""
    store = _pending_store()
    tenant = _install_tenant()

    row = store.get(provenance_id, tenant)
    if row is None:
        raise HTTPException(status_code=404, detail={"error": "no_pending_reapproval"})

    from yashigani.auth.stepup import (
        PrivilegedMutationContext,
        assert_privileged_mutation,
    )

    ctx = PrivilegedMutationContext(
        reason="mcp.envelope.reject",
        principal=session.account_id,
        target=provenance_id,
        before={"triage_class": row.get("triage_class", "")},
        after={"decision": "keep_blocked"},
    )
    # Raises StepUpRequired (401) / NotAuthorised (403) on failure → JS prompts.
    assert_privileged_mutation(session, ctx, audit_writer=backoffice_state.audit_writer)

    store.resolve(provenance_id, tenant)
    logger.warning(
        "capability-envelope KEEP-BLOCKED via admin UI: provenance=%.12s by=%s "
        "(block remains latched; MCP stays gated)",
        provenance_id, session.account_id,
    )
    return {
        "status": "ok",
        "provenance_id": provenance_id,
        "message": "Mutation rejected — the MCP stays blocked until an explicit re-approval.",
    }

File: src/yashigani/backoffice/static/js/envelope_reapproval.js
// Yashigani Backoffice — MCP capability-envelope RE-APPROVAL admin panel (3.0)
// Thin consumer of /admin/mcp/envelopes/* — the operator-facing surface for the
// imported-MCP capability-envelope pin (YSG-RISK-060).
//
// Depends on: dashboard.js (api, apiMutate, escapeHtml, the step-up modal)
// Loaded as: <script src="/static/js/envelope_reapproval.js" defer></script>
//
// SECURITY — attacker-controlled content escaping (CWE-79):
//   Tool keys (provenance_id::tool_name) and every diff `detail` string come
//   from an UNTRUSTED upstream MCP's tools/list — this is the XSS surface the
//   whole feature exists to police, so a malicious tool name like
//   `<img src=x onerror=...>` MUST render inert.  EVERY value rendered into
//   innerHTML below passes through escapeHtml().  No value from the server is
//   ever interpolated raw.
//
//   No secret, handle, or map material is ever returned by these endpoints, so
//   nothing sensitive reaches the DOM — the panel renders only the public,
//   typed capability dimensions + the diff details.
//
// Step-up: approve/reject go through apiMutate(), which transparently catches
// the 401 `step_up_required` from the server-side privileged-mutation gate,
// shows the shared TOTP modal, and retries — no bespoke step-up code here.
//
// Last updated: 2026-06-10

/* global api, apiMutate, escapeHtml */

// The provenance_id of the entry currently open in the diff panel.
var _envSelected = null;

var ENV_SEVERITY_BADGE = { high: 'badge-red', med: 'badge-amber', low: 'badge-blue' };
var ENV_REASON_LABEL = {
    expanding: 'Capability expansion',
    uncertain: 'Sidecar uncertain (fail-closed)'
};

// ---------------------------------------------------------------------------
// Load: the pending queue
// ---------------------------------------------------------------------------

async function loadEnvelopes() {
    _envHideDiff();
    var tbody = document.getElementById('env-pending-tbody');
    var resultEl = document.getElementById('env-pending-result');
    if (!tbody) { return; }
    tbody.innerHTML = '<tr><td colspan="7" class="empty">Loading…</td></tr>';
    if (resultEl) { resultEl.innerHTML = ''; }

    var data = await api('/admin/mcp/envelopes/pending');
    if (!data || !data.pending) {
        tbody.innerHTML = '<tr><td colspan="7" class="empty">Failed to load pending re-approvals.</td></tr>';
        return;
    }
    if (!data.pending.length) {
        tbody.innerHTML = '<tr><td colspan="7" class="empty">No pending re-approvals — every imported MCP is within its approved envelope.</td></tr>';
        return;
    }
    var html = '';
    data.pending.forEach(function (p) {
        var reasonBadge = (p.triage_class === 'expanding') ? 'badge-red' : 'badge-amber';
        var reasonText = ENV_REASON_LABEL[p.triage_class] || p.triage_class;
        var prov = p.provenance_id || '';
        var when = p.blocked_at ? new Date(p.blocked_at * 1000).toISOString().slice(0, 19).replace('T', ' ') : '—';
        html += '<tr>'
            + '<td>' + escapeHtml(p.server_id) + '</td>'
            + '<td><code>' + escapeHtml(prov.slice(0, 12)) + '…</code></td>'
            + '<td><span class="badge ' + reasonBadge + '">' + escapeHtml(reasonText) + '</span></td>'
            + '<td>' + escapeHtml(String(p.candidate_tool_count)) + '</td>'
            + '<td><code>' + escapeHtml(p.candidate_egress_posture) + '</code></td>'
            + '<td>' + escapeHtml(when) + '</td>'
            // data-prov carries the (server-supplied, untrusted) id — we escape it
            // for the attribute AND only ever use it via encodeURIComponent in fetch.
            + '<td><button class="btn-sm-save" data-action="envReview" data-prov="' + escapeHtml(prov) + '">Review</button></td>'
            + '</tr>';
    });
    tbody.innerHTML = html;
}

// ---------------------------------------------------------------------------
// Review: the field-level diff vs ORIGINAL baseline (+ vs prior)
// ---------------------------------------------------------------------------

async function envReview(provenanceId) {
    _envSelected = provenanceId;
    var emptyEl = document.getElementById('env-diff-empty');
    var bodyEl = document.getElementById('env-diff-body');
    var decEl = document.getElementById('env-decision-result');
    if (decEl) { decEl.innerHTML = ''; }
    if (emptyEl) { emptyEl.className = ''; emptyEl.textContent = 'Loading change review…'; }
    if (bodyEl) { bodyEl.className = 'env-hidden'; }

    var data = await api('/admin/mcp/envelopes/pending/' + encodeURIComponent(provenanceId));
    if (!data) {
        if (emptyEl) { emptyEl.className = ''; emptyEl.innerHTML = '<span class="badge badge-red">Could not load the diff (the entry may have been resolved). Refresh the queue.</span>'; }
        return;
    }
    _envRenderDiff(data);
}

function _envRenderDiff(data) {
    var emptyEl = document.getElementById('env-diff-empty');
    var bodyEl = document.getElementById('env-diff-body');
    if (emptyEl) { emptyEl.className = 'env-hidden'; }
    if (bodyEl) { bodyEl.className = ''; }

    // Headline — the legible "tried to expand from X to Y" story.
    var head = document.getElementById('env-diff-headline');
    if (head) {
        var reasonText = ENV_REASON_LABEL[data.triage_class] || data.triage_class;
        var s = '<div class="env-headline-box">'
            + '<span class="badge badge-red">BLOCKED</span> '
            + '<strong>' + escapeHtml(data.server_id) + '</strong> '
            + '<span class="txt-note">(' + escapeHtml(reasonText) + ')</span><br>';
        var origTools = (data.original && data.original.tool_count) || 0;
        var candTools = (data.candidate && data.candidate.tool_count) || 0;
        s += 'Originally approved: <strong>' + escapeHtml(String(origTools)) + '</strong> tool(s), egress <code>'
            + escapeHtml((data.original || {}).egress_posture || 'NONE') + '</code>. '
            + 'Now requesting: <strong>' + escapeHtml(String(candTools)) + '</strong> tool(s), egress <code>'
            + escapeHtml((data.candidate || {}).egress_posture || 'NONE') + '</code>.';
        if (data.egress_change) {
            s += '<br><span class="badge badge-red">Egress posture change</span> '
                + '<code>' + escapeHtml(data.egress_from) + '</code> &rarr; <code>' + escapeHtml(data.egress_to) + '</code> '
                + '<span class="txt-note">(this MCP wants to reach further than it was approved for)</span>';
        }
        s += '</div>';
        head.innerHTML = s;
    }

    _envFillFindings('env-vsorig-tbody', data.vs_original, 'No capability expansion vs the original — within the approved envelope.');
    _envFillFindings('env-vsprior-tbody', data.vs_prior, 'No change vs the immediately-prior approved version.');
}

function _envFillFindings(tbodyId, findings, emptyMsg) {
    var tbody = document.getElementById(tbodyId);
    if (!tbody) { return; }
    findings = findings || [];
    if (!findings.length) {
        tbody.innerHTML = '<tr><td colspan="3" class="empty">' + escapeHtml(emptyMsg) + '</td></tr>';
        return;
    }
    var html = '';
    findings.forEach(function (f) {
        // dimension label + every string is attacker-influenced → escapeHtml.
        var sevBadge = ENV_SEVERITY_BADGE[f.severity] || 'badge-p4';
        var tool = f.tool_key ? ('<code>' + escapeHtml(f.tool_key) + '</code>') : '<span class="txt-note">—</span>';
        html += '<tr>'
            + '<td><span class="badge ' + sevBadge + '">' + escapeHtml(f.label || f.dimension) + '</span></td>'
            + '<td>' + tool + '</td>'
            + '<td>' + escapeHtml(f.detail) + '</td>'
            + '</tr>';
    });
    tbody.innerHTML = html;
}

function _envHideDiff() {
    _envSelected = null;
    var emptyEl = document.getElementById('env-diff-empty');
    var bodyEl = document.getElementById('env-diff-body');
    if (emptyEl) { emptyEl.className = ''; emptyEl.textContent = 'Select a pending re-approval above to review the field-level diff.'; }
    if (bodyEl) { bodyEl.className = 'env-hidden'; }
}

// ---------------------------------------------------------------------------
// Approve / Reject — both step-up gated server-side (apiMutate triggers modal)
// ---------------------------------------------------------------------------

async function envApprove() {
    if (!_envSelected) { return; }
    if (!confirm('Re-approve this tool-surface change? This re-pins a NEW baseline and clears the block. Requires TOTP step-up.')) { return; }
    await _envDecision('approve', 'Re-approved — new baseline pinned, block cleared.');
}

async function envReject() {
    if (!_envSelected) { return; }
    if (!confirm('Keep this MCP BLOCKED? The mutation is rejected and the MCP stays gated until an explicit re-approval. Requires TOTP step-up.')) { return; }
    await _envDecision('reject', 'Rejected — the MCP stays blocked.');
}

async function _envDecision(action, okMsg) {
    var prov = _envSelected;
    var decEl = document.getElementById('env-decision-result');
    if (decEl) { decEl.innerHTML = '<span class="loading">Submitting…</span>'; }

    // apiMutate() catches the server's 401 step_up_required, shows the TOTP modal,
    // and retries — so the REAL privileged-mutation gate is enforced server-side.
    var resp = await apiMutate(
        '/admin/mcp/envelopes/pending/' + encodeURIComponent(prov) + '/' + action,
        { method: 'POST' }
    ).catch(function () { return null; });

    if (!resp) {
        if (decEl) { decEl.innerHTML = '<span class="badge badge-red">Network error or step-up cancelled.</span>'; }
        return;
    }
    if (!resp.ok) {
        var err = await resp.json().catch(function () { return {}; });
        var msg = (err.detail && (err.detail.error || err.detail.message)) ? (err.detail.error || err.detail.message) : ('HTTP ' + resp.status);
        if (resp.status === 403) { msg = 'Forbidden — operator (admin) privilege required.'; }
        if (decEl) { decEl.innerHTML = '<span class="badge badge-red">' + escapeHtml(msg) + '</span>'; }
        return;
    }
    if (decEl) { decEl.innerHTML = '<span class="badge badge-green">' + escapeHtml(okMsg) + '</span>'; }
    _envHideDiff();
    loadEnvelopes();
}

// ---------------------------------------------------------------------------
// Self-register action handlers (CSP-safe — no inline handlers)
// ---------------------------------------------------------------------------

document.addEventListener('click', function (e) {
    var actionEl = e.target.closest ? e.target.closest('[data-action]') : null;
    if (!actionEl) { return; }
    switch (actionEl.getAttribute('data-action')) {
        case 'envRefresh':
            loadEnvelopes();
            break;
        case 'envReview':
            envReview(actionEl.getAttribute('data-prov'));
            break;
        case 'envApprove':
            envApprove();
            break;
        case 'envReject':
            envReject();
            break;
        default:
            break;
    }
});

// Expose for showPage() in dashboard.js
window.loadEnvelopes = loadEnvelopes;
